Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -2461,6 +2461,13 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
     if (completedURL.protocolIsJavaScript() && frameElement() && !frameElement()->document().contentSecurityPolicy()->allowJavaScriptURLs(aboutBlankURL().string(), { }, completedURL.string(), frameElement()))
         return;
 
+    RefPtr mainFrame = dynamicDowncast<LocalFrame>(frame->mainFrame());
```

Comment:
I would think we want the parent frame rather than the main frame.  If a.com has an iframe with b.com that has an iframe with c.com, then a.com didn't "refer" to c.com.  b.com did, right?

Reply:
Anne seems to disagree.  He's probably right, but I'm not sure.

Reply:
Your scenario is for the initial navigation which is based on the attribute value. That's different. In this case it's really about who called the method and there's no reason for it to be different from the other `Location` methods. (Which in part is why I argued for refactoring it so they all invoke the same primitive.)

Reply:
OK, further testing indicates that the parent frame is actually what Gecko and Blink set the Referer from.

So, because the purpose of this change has, so far, been to achieve interoperability/compat with Gecko and Blink, I’ve updated the implementation to use the parent frame rather than the “main” frame — and I’ve added more tests.

> If a.com has an iframe with b.com that has an iframe with c.com, then a.com didn't "refer" to c.com. b.com did, right?

Right it seems — yes, as far as what happens when script running in b.com calls `location.replace` or  `location.assign` are called on a c.com iframe.

Specifically:
1. The `assign-with-nested-iframe.html` and `replace-with-nested-iframe.html` tests I added are the a.com top-level documents that have a b.com iframe (`resources/replace-or-assign-call-on-iframe.html`) that in turn also has an iframe (c.com).
2. Script running in the b.com (`resources/replace-or-assign-call-on-iframe.html`) iframe calls  `location.replace` or  `location.assign` on the c.com iframe
3. The Referer for the c.com request in Gecko and Blink is the parent b.com (`resources/replace-or-assign-call-on-iframe.html`) iframe — _not_ the top-level `assign-with-nested-iframe.html` and `replace-with-nested-iframe.html` documents (“main” frames).

It’s not clear to me which part of the spec states what the required behavior is for that case — and not clear to me if Gecko and Blink behavior is violating the relevant spec requirements.

But if we want to have interop/compat with Gecko and Blink on this, then it seems we need to use the parent frame as the referrer in the above case. And from the spec side, if it seems desirable to have the spec match the existing Gecko and Blink behavior, and the spec needs to change to do that — then I’d be happy to also work on spec patch.

If on the other hand, the existing Gecko and Blink behavior here seems undesirable — if it seems we don’t want WebKit to match that — then I could change the implementation here back to using whatever the current spec actually requires (which is maybe the “main” window, but I’m still actually not clear if that’s what the spec says…).

And in that case, I guess I’d also need to raise Gecko and Blink bugs to try to get them to change (again, with the goal still being to get interop on this).